factor/grace/classes/factor.php

```diff
@@ -0,0 +1,158 @@
+
+use tool_mfa\local\factor\object_factor_base;
+
+class factor extends object_factor_base {
+
+    /**
+     * Grace Factor implementation.
+     * This factor needs no user setup, return true.
+     *
+     * {@inheritDoc}
+     */
+    public function setup_user_factor($data) {
+        return true;
+    }
+
+    /**
+     * Grace Factor implementation.
+     * This factor is a singleton, return single instance.
+     *
+     * {@inheritDoc}
+     */
+    public function get_all_user_factors() {
+        global $DB, $USER;
+
+        if (!$DB->record_exists('factor_grace', array('userid' => $USER->id))) {
+            $DB->insert_record('factor_grace', array(
+                'userid' => $USER->id,
+                'ip' => $USER->lastip,
+                'timecreated' => time(),
+            ));
+        }
+        $factorrecord = $DB->get_record('factor_grace', array('userid' => $USER->id));
+
+        $factor = (object) array(
+            'id' => 1,
+            'name' => $this->name,
+            'devicename' => '-',
+            'timecreated' => $factorrecord->timecreated,
+            'createdfromip' => $factorrecord->timecreated,
+            'lastverified' => '-',
+            'revoked' => '-'
+        );
+
+        return [$factor];
+    }
+
+    /**
+     * Grace Factor implementation.
+     * Factor has no input.
+     *
+     * {@inheritDoc}
+     */
+    public function has_input() {
+        return false;
+    }
+
+    /**
+     * Grace Factor implementation.
+     * Checks the user login time against their first login after MFA activation.
+     *
+     * {@inheritDoc}
+     */
+    public function get_state() {
+        // Check if user already has a record, if not, create one at current time.
+        global $USER, $DB;
+
+        if (!$DB->record_exists('factor_grace', array('userid' => $USER->id))) {
+            $DB->insert_record('factor_grace', array(
+                'userid' => $USER->id,
+                'ip' => $USER->lastip,
+                'timecreated' => time(),
+            ));
+        }
+        $record = $DB->get_record('factor_grace', array('userid' => $USER->id));
+
+        $starttime = $record->timecreated;
+
+        // If no start time is recorded, status is unknown.
+        if (empty($starttime)) {
+            return \tool_mfa\plugininfo\factor::STATE_UNKNOWN;
+        } else {
+            $duration = get_config('factor_grace', 'graceperiod');
+
+            if (!empty($duration)) {
+                return (time() <= $starttime + $duration)
+                ? \tool_mfa\plugininfo\factor::STATE_PASS
+                : \tool_mfa\plugininfo\factor::STATE_NEUTRAL;
+            } else {
+                return \tool_mfa\plugininfo\factor::STATE_UNKNOWN;
+            }
+        }
+    }
+
+    /**
+     * Grace Factor implementation.
+     * State cannot be set. Return true.
+     *
+     * {@inheritDoc}
+     */
+    public function set_state($state) {
+        return true;
```

Comment:
Not that we are actually using the return value at all, but typically most api's return true if it was set. Made https://github.com/catalyst/moodle-tool_mfa/issues/95 to handle this